### DISCUSSION TITLE
[Native DLL] Should I remove all `const` in my translation?

### DISCUSSION THREAD

**Discussion by ghost** [pascalabcnet/pascalabcnet#3024] [General]
The C API is full of `const`. I know that if I removed them in my translation, everything still works. But should I do so?

**Comment by SunSerega** [comment]
`const` in the native API is a promise about the memory usage by that API.

---

The simplest example with strings:
```
void f1(char const* s)
void f2(char* s)
```
The `*` divides the parameter into 2 levels of indirection:
1. The variable `s`
2. The first char in the string, which `s` points to

Since there is no `const` to the right of `*` - it means the native function reserves for itself the right to do `s = "ABC"`. Which is meaningless to you - it overrides the parameter variable but doesn't change the characters of the string you passed in. That's why it's usually omitted from the headers visible to you - this `const` wouldn't tell you anything.

Now, for the actually meaningful part: `char const` (or `const char`, the order doesn't matter in the text block split by `*`)
Here `const` tells you that the library promises that it won't change the chars your string points to. And so I'd expect this parameter can be decorated with `[In]`, to explicitly tell the default marshaler to only convert your string one way. Tho AFAIK this is already the default for strings (but not arrays!).

The non-`const` overload is more complicated. Since there is no promise - it's probably because the function actually needs to write to this string. But it still could be `[In, Out]` or just `[Out]`.
Or, maybe the API authors forgot `const`. And maybe they couldn't add it later. I'm not joking, I was caught off guard by that myself. For instance, look at [`GetPerfQueryIdByNameINTEL`](https://registry.khronos.org/OpenGL/extensions/INTEL/INTEL_performance_query.txt). Parameter `queryName` is supposed to only take the `[In]` string, but initially my code-generators assumed not having `const` means it's an `[Out]`-only parameter.

Likewise, having `const` doesn't mean the function will never modify the contents of the string - since it's just a promise, and an easy one to break in C.

---

There can often also be stuff like:
```
void f(const char*** s)
```
Basically, an `array of array of string`.

Just based on `const`:
```
s = ...; // Native code can do, but you wouldn't notice that
s[0] = ...; // Can set, you'll see that one of the "array of string"-s has been entirely replaced, with a bunch of new strings
s[0][0] = ...; // Can set, you'll see that one of the "string"-s has been replaced
s[0][0][0] = ...; // Cannot replace any character in an existing string
```
If the API says it can replace some strings and arrays - this would be an actual hell to marshal correctly.
But would API do something like that? Probably not, that'd be pretty stupid.
More likely people didn't want to write the more correct but ugly:
```
void f(char const *const *const *s)
```

---


That is all to say: Don't rely on `const` modifiers but do look at them.
When you manually write `external` functions (without code-gen) it's better to rely on text documentation and intuition.

---

On another note, have you figured out yet how to marshal `[Out]` strings? Because just adding the `[Out]` attribute wouldn't change the fact that you copy reference to the string when you pass it in.
There are 2 main ways. The easy one and the efficient one.
If not - try to find them first, it might take an even bigger post to explain that from 0...
